ExpressMapper NET40/MappingServiceBase.cs

```diff
@@ -423,28 +412,56 @@ protected LoopExpression CreateLoopExpression(Expression doMoveNextSrc, BlockExp
             {
                 return Expression.Call(destColl, destList.GetMethod("ToArray"));
             }
-            if (!destPropType.IsGenericType) return destColl;
+
+            if (!destPropType.IsGenericType && GetEnumerableInterface(destPropType) == null)
+            {
+                return destColl;
+            }
 
             if (typeof(IQueryable).IsAssignableFrom(destPropType))
             {
                 return Expression.Call(typeof(Queryable), "AsQueryable", new[] { destType }, destColl);
             }
 
-            if (destPropType.IsInterface && destColl.Type.IsSubclassOf(destPropType))
-            {
+            if (destPropType.IsInterface && destPropType.IsAssignableFrom(destColl.Type)) {
+                // This will handle any destination interface implemented by List<T>
                 return destColl;
             }
 
-            if (destPropType.IsClass)
-            {
+            ConstructorInfo ctor = null;
 
+            if (destPropType.IsInterface)
+            {
+                // We are targeting an interface type, we need to find a compatible collection type
+                // We could look for a loaded type that implements the target interface and has an appropriate
+                // constructor, but that is a bit too much magic for now.
+                throw new NotImplementedException(
+                    "Destination interface type " + destPropType.FullName + " is not supported yet" );

```

Comment:
I did not implement the case where the destination collection type is an interface that is not implemented by List.